googlemock/include/gmock/gmock-generated-function-mockers.h.pump

```diff
@@ -140,24 +148,34 @@ $var as = [[$for j, [[gmock_a$j]]]]
 $var matcher_as = [[$for j, \
                      [[GMOCK_MATCHER_(tn, $j, __VA_ARGS__) gmock_a$j]]]]
 // INTERNAL IMPLEMENTATION - DON'T USE IN USER CODE!!!
-#define GMOCK_METHOD$i[[]]_(tn, constness, ct, Method, ...) \
+#define GMOCK_METHOD$i[[]]_(tn, qualifiers, ct, Method, ...) \
   GMOCK_RESULT_(tn, __VA_ARGS__) ct Method( \
-      $arg_as) constness { \
+      $arg_as) qualifiers { \
     GTEST_COMPILE_ASSERT_((::testing::tuple_size<                          \
         tn ::testing::internal::Function<__VA_ARGS__>::ArgumentTuple>::value == $i), \
         this_method_does_not_take_$i[[]]_argument[[$if i != 1 [[s]]]]); \
-    GMOCK_MOCKER_($i, constness, Method).SetOwnerAndName(this, #Method); \
-    return GMOCK_MOCKER_($i, constness, Method).Invoke($as); \
+    ::testing::internal::RemoveVolatileCast(*this) \
+        .GMOCK_MOCKER_($i, Method) \
+        .SetOwnerAndName(& ::testing::internal::RemoveVolatileCast(*this), #Method); \
+    return ::testing::internal::RemoveVolatileCast(*this) \
+        .GMOCK_MOCKER_($i, Method) \
+        .Invoke($as); \
   } \
   ::testing::MockSpec<__VA_ARGS__>& \
-      gmock_##Method($matcher_as) constness { \
-    GMOCK_MOCKER_($i, constness, Method).RegisterOwner(this); \
-    return GMOCK_MOCKER_($i, constness, Method).With($as); \
+      gmock_##Method($matcher_as) qualifiers { \
+    ::testing::internal::RemoveVolatileCast(*this) \
+        .GMOCK_MOCKER_($i, Method) \
+        .RegisterOwner(& ::testing::internal::RemoveVolatileCast(*this); \
+    return ::testing::internal::RemoveVolatileCast(*this) \
```

Comment:
To be honest, as I am neither the original author of these patches (I just applied them, fixed some obvious patch-errors and provided the pull-request; see my note in the first message and the referenced issue ticket) nor have I originally written the macros that get patched, I cannot tell for sure.

What I understand is that the macros `GMOCK_METHOD$i` are the macros that get called by the public macros `MOCK_METHOD$i` (and the new `MOCK_QUALIFIED_METHOD$i` macros).
`GMOCK_METHOD$i` then calls `GMOCK_MOCKER_($i, Method)` which just creates the name of the member-object which is the mocker for the particular function.
And to be able to call the required non-volatile member-function (`SetOwnerAndName`, `RegisterOwner`, `Invoke` etc.) the volatile qualifier must be casted away.

If the `this` object was not created as volatile in the first place, the volatile qualifier just got added by calling the (mocked) volatile function and can be safely casted away.
However, you are right, at this point we cannot tell where the volatile qualifier was added to `this`. Therefore we might trigger undefined behavior.

But is it possible (or useful at all) to create a volatile mock-object? (I cannot think of any useful use-case.)
If not, we could probably safely assume that volatile just got added by calling the (mocked) function and can therefore by casted away without triggering undefined behavior.